### PR TITLE
fix: don't enforce lwc:dom=manual restrictions on light DOM components

### PR DESCRIPTION
**Pull request by nolanlawson**
## Details

`lwc:dom="manual"` (aka portal) restrictions should only apply to shadow DOM, not light DOM.

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

## GUS work item

W-9473311

### CODE REVIEW

File: packages/integration-karma/test/light-dom/restriction/index.spec.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    });
+
+    it('should restrict setting outerHTML on light DOM component', () => {
+        const elm = createElement('x-component', { is: Component });
+        document.body.appendChild(elm);
+
+        expect(() => {
+            elm.querySelector('div').outerHTML = 'foo';
+        }).toThrowError(TypeError, 'Invalid attempt to set outerHTML on Element.');
+    });
```
Comment (nolanlawson) [inline packages/integration-karma/test/light-dom/restriction/index.spec.js:30]: `patchElementWithRestrictions` also restricts setting `outerHTML`. I don't know what purpose this serves, but it seems unrelated to portals.

File: packages/integration-karma/test/shadow-dom/Element-properties/Element.innerHTML.spec.js [hunk 2]
```diff
@@ -38,16 +38,18 @@ describe('Element.innerHTML - set', () => {
...
-
-        const div = elm.shadowRoot.querySelector('div');
-
-        expect(() => {
-            div.innerHTML = '<span>Hello World!</span>';
-        }).toLogErrorDev(
-            /\[LWC error\]: The `innerHTML` property is available only on elements that use the `lwc:dom="manual"` directive./
-        );
-    });
+    if (!process.env.NATIVE_SHADOW) {
```
Comment (pmdartus) [inline packages/integration-karma/test/shadow-dom/Element-properties/Element.innerHTML.spec.js:41]: I don't think this is the appropriate approach to test this. With this change, the test only runs when synthetic shadow is enabled. 

Instead, we should test that the error message is only reported with synthetic shadow DOM.

```js
const invokeInnerHTML = () => div.innerHTML = '<span>Hello World!</span>';

if (process.env.NATIVE_SHADOW) {
  expect(invokeInnerHTML).not.toLogErrorDev();
} else {
  expect(invokeInnerHTML).toLogErrorDev(/\[LWC error\]: The `innerHTML` property is available only on elements that use the `lwc:dom="manual"` directive./);
}
```

File: packages/integration-karma/test/template/directive-lwc-dom-manual/index.spec.js [hunk 3]
```diff
@@ -13,43 +13,45 @@ function waitForStyleToBeApplied() {
...
-            const root = createElement('x-without-lwc-dom-manual', { is: withoutLwcDomManual });
-            document.body.appendChild(root);
-            const elm = root.shadowRoot.querySelector('div');
-
-            expect(() => fn(elm)).toLogErrorDev(
-                new RegExp(
-                    `\\[LWC error\\]: The \`${method}\` method is available only on elements that use the \`lwc:dom="manual"\` directive.`
-                )
-            );
+if (!process.env.NATIVE_SHADOW) {
```
Comment (pmdartus) [inline packages/integration-karma/test/template/directive-lwc-dom-manual/index.spec.js:16]: Ditto.

File: packages/integration-karma/test/shadow-dom/Node-properties/Node.textContent.spec.js [hunk 4]
```diff
@@ -37,16 +37,18 @@ describe('Node.textContent - setter', () => {
...
-
-        const div = elm.shadowRoot.querySelector('div');
-
-        expect(() => {
-            div.textContent = '<span>Hello World!</span>';
-        }).toLogErrorDev(
-            /\[LWC error\]: The `textContent` property is available only on elements that use the `lwc:dom="manual"` directive./
-        );
-    });
+    if (!process.env.NATIVE_SHADOW) {
```
Comment (pmdartus) [inline packages/integration-karma/test/shadow-dom/Node-properties/Node.textContent.spec.js:40]: Ditto

File: packages/integration-karma/test/light-dom/restriction/index.spec.js (hunk 1 above)
Comment (pmdartus) [reply packages/integration-karma/test/light-dom/restriction/index.spec.js:30]: I also don't think it is related to portals. The goal is to avoid that components get unounted using `outHTML`. That said it is quite an edge case.

**Review by pmdartus — CHANGES_REQUESTED** [review]

File: packages/integration-karma/test/shadow-dom/Element-properties/Element.innerHTML.spec.js (hunk 2 above)
Comment (nolanlawson) [reply packages/integration-karma/test/shadow-dom/Element-properties/Element.innerHTML.spec.js]: I updated this PR to only be about light DOM, because that's what the original bug was about, and @ekashida pointed out that we don't necessarily want to encourage component authors to deliberately break compat with synthetic shadow.

**Review by pmdartus — APPROVED** [review]

**Review by ekashida — APPROVED** [review]

**Review by abdulsattar — APPROVED** [review]

File: packages/@lwc/engine-core/src/framework/restrictions.ts [hunk 5]
```diff
@@ -91,7 +94,7 @@ export function patchElementWithRestrictions(elm: Element, options: { isPortal:
     };
 
     // Apply extra restriction related to DOM manipulation if the element is not a portal.
-    if (isFalse(options.isPortal)) {
+    if (isFalse(options.isLight) && isFalse(options.isPortal)) {
```
Comment (ekashida) [inline packages/@lwc/engine-core/src/framework/restrictions.ts:97]: Nit: We agreed in the past that `isTrue/isFalse` should only be used in place of expressions like `foo === true` or `bar === false`. Since `isLight` and `isPortal` are boolean-typed and we don't usually use `===` when we know the thing is a boolean, we wouldn't have used this utility function here.

Than being said, I don't think we've been following this guideline consistently so I'm fine either way :)

Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/restrictions.ts:97]: I'm happy to replace this with just `!options.isLight && !options.isPortal`. And thanks for explaining what `isTrue`/`isFalse` did; I didn't quite get it before. :)
